### PR TITLE
Run runtime compilation in its own module

### PR DESCRIPTION
**Pull request by ISauve**
### What does this PR do?

Creates a new module in the system-probe which is responsible for performing runtime compilation on all the eBPF programs which require it.

This module will run first, so that all eBPF programs are compiled prior to any other module starting. Then, when a program is required, the compiled output can simply be read from the filesystem.

The runtime compilation module will be enabled if runtime compilation is enabled for any one (or more) eBPF programs. See additional notes for details on how to enable all our current runtime-compilable eBPF programs.

### Motivation

In containerized environments, runtime compilation and kernel header downloading are a known source of memory problems in the system-probe. Although a lot of progress has been made in alleviating the memory impact of both processes, having the ability to perform runtime compilation & kernel header downloading as an independent step (i.e. via the new `compile` command) makes it possible for  runtime compilation to occur in a kubernetes `init` container prior to the system-probe container running. This removes the memory impact on the system-probe container entirely.

In a non-containerized environment, having the ability to run the `compile` command independently will make it easier to debug runtime compilation-related issues, as well as grant increased flexibility to anyone who wishes to enable runtime compilation in their system.

In addition, in its current state the configuration of runtime compilation is unclear. There are runtime compilable assets in all the modules of the system-probe, and if you disable runtime compilation in the system-probe configuration it can still be enabled someplace else. By organizing all the runtime-compilation related logic into a single module (with a configuration option to allow skipping the whole module) and compiling all the eBPF programs sequentially, it becomes clear exactly which runtime compiled features are enabled/disabled, and runtime compilation as a whole is simpler to configure.

### Additional Notes

This PR slightly changes the way runtime compilation is configured. The new configuration is as follows:
```
system_probe_config:
  enable_tcp_queue_length: true # enables runtime compilation of oom-kill.c
  enable_oom_kill: true         # enables runtime compilation of  tcp-queue-length.c

network_config:
  enabled: true 
  enable_runtime_compilation: true # enables runtime compilation of tracer.c & conntrack.c
  enable_http_monitoring: true     # enables runtime compilation of http.c

runtime_security_config:
   enabled: true
   runtime_compilation:
     enabled: true                     # enables runtime compilation of runtime-security.c
     compiled_constants_enabled: true  # enables runtime compilation of constant_fetcher.c

# Runtime-compilation settings now live under their own section
runtime_compilation_config:
  runtime_compiler_output_dir:
  enable_kernel_header_download:
  kernel_header_dirs:
  kernel_header_download_dir:
  apt_config_dir:
  yum_repos_dir:
  zypper_repos_dir:
  skip_compilation: 
```
Note that the global setting for disabling runtime compilation has been changed from `system_probe_config.enable_runtime_compiler` to `runtime_compilation_config.skip_compilation`. If this configuration option is used, the runtime compilation module will be skipped & compilation will not occur - though runtime compiled assets will still be loaded wherever runtime compilation is enabled. (Previously, even if `enable_runtime_compiler` was set to `false`, it was possible for runtime compilation to occur anyways - i.e. if the tcp queue length checks or oom kill checks were enabled, or if runtime compiled constants were not explicitly disabled and the running architecture is COS).

### Describe how to test/QA your changes

1) Enable all the runtime-compilable assets in your `system-probe.yaml`:
```
system_probe_config:
  enable_tcp_queue_length: true
  enable_oom_kill: true

network_config:
  enabled: true
  enable_runtime_compilation: true
  enable_http_monitoring: true

runtime_security_config:
  enabled: true
  runtime_compilation:
    enabled: true
    compiled_constants_enabled: true
```
2) Run the system-probe's `compile` command and ensure that all assets compile properly:
```
| Starting the runtime compiler
| successfully compiled runtime version of tracer.c
| successfully compiled runtime version of conntrack.c
| successfully compiled runtime version of http.c
| successfully compiled runtime version of runtime-security.c
| successfully compiled runtime version of constant_fetcher.c
| successfully compiled runtime version of tcp-queue-length.c
| successfully compiled runtime version of oom-kill.c
| Runtime compilation complete
```
3) Run the system-probe's  `compile` command again and ensure the previously compiled assets are all found:
```
| Starting the runtime compiler
| found previously compiled runtime version of tracer.c
| found previously compiled runtime version of conntrack.c
| found previously compiled runtime version of http.c
| found previously compiled runtime version runtime-security.c
| found previously compiled runtime version of constant_fetcher.c
| found previously compiled runtime version of tcp-queue-length.c
| found previously compiled runtime version of oom-kill.c
| Runtime compilation complete
```
4) Delete all the files in your runtime compiler output directory (by default, this is `/var/tmp/datadog-agent/system-probe/build`)
5) Run the system-probe normally. Once again, check for log lines indicating that all assets compiled successfully:
```
| Starting the runtime compilation module
| successfully compiled runtime version of tracer.c
| successfully compiled runtime version of conntrack.c
| successfully compiled runtime version of http.c
| successfully compiled runtime version of runtime-security.c
| successfully compiled runtime version of constant_fetcher.c
| successfully compiled runtime version of tcp-queue-length.c
| successfully compiled runtime version of oom-kill.c
| Runtime compilation module completed successfully
```
6) Read through the logs and make sure that you don't see any log lines indicating that a runtime compiled asset could not be loaded (i.e. a log line containing "failed to fetch compiled [asset name]").
7) In a separate window, query the system-probe for runtime compilation telemetry and verify that it looks correct:
```
$ curl -s --unix-socket /opt/datadog-agent/run/sysprobe.sock http://unix/connections?client_id=1 | jq .compilationTelemetryByAsset
{
  "conntrack": {
    "runtimeCompilationEnabled": true,
    "runtimeCompilationResult": "CompiledOutputFound",
    "runtimeCompilationDuration": "444002"
  },
  "constant_fetcher": {
    "runtimeCompilationEnabled": true,
    "runtimeCompilationResult": "CompiledOutputFound",
    "runtimeCompilationDuration": "84386"
  },
  "http_tracer": {
    "runtimeCompilationEnabled": true,
    "runtimeCompilationResult": "CompiledOutputFound",
    "runtimeCompilationDuration": "597059"
  },
  (etc...)
}
```
8) Query the system-probe for kernel header downloading telemetry and verify that it looks correct too:
```
$ curl -s --unix-socket /opt/datadog-agent/run/sysprobe.sock http://unix/connections?client_id=1 | jq .kernelHeaderFetchResult
"DefaultHeadersFound"
```

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/runtimecompiler/config/config.go
Comment (vickenty) [inline pkg/runtimecompiler/config/config.go:73]: I think options are swapped here?

**Review by vickenty — COMMENTED** [review]
Please set CODEOWNERS for the new package, otherwise I think there is nothing here for agent-core to review.

File: pkg/runtimecompiler/config/config.go
Comment (ISauve) [reply pkg/runtimecompiler/config/config.go:73]: Yep, good catch!

**Review by vickenty — APPROVED** [review]

File: cmd/system-probe/app/compile.go
Comment (hmahmood) [inline cmd/system-probe/app/compile.go:44]: The concern here is that any container wouldn't start right, including the `agent`? And `system-probe` would start with an error presumably if compilation fails?

File: cmd/system-probe/modules/compile.go
Comment (hmahmood) [inline cmd/system-probe/modules/compile.go:59]: No stats?

File: pkg/config/system_probe.go
Comment (hmahmood) [inline pkg/config/system_probe.go:171]: How do `enable_runtime_compliation` and `skip_runtime_compilation` relate? Why do we need this one?

File: pkg/ebpf/bytecode/runtime/runtime_compilation_helpers.go
Comment (hmahmood) [inline pkg/ebpf/bytecode/runtime/runtime_compilation_helpers.go:110]: Can't we just do
```go
hasher := sha256.New()
for _, d := range defaultFlags {
    hasher.Write([]byte(d))
}
for _, a := range additionalFlags {
    hasher.Write([]byte(a))
}
```

File: pkg/ebpf/bytecode/runtime/runtime_compilation_helpers.go
Comment (hmahmood) [inline pkg/ebpf/bytecode/runtime/runtime_compilation_helpers.go:165]: Not sure why `string` is needed here
```go
sha256hex(append(uname.Release, uname.Version...))
```
We could also have 
```go
func sha256hex(bufs ...[]byte)
```
... and then we don't have to use `append`.

File: pkg/runtimecompiler/config/config.go
Comment (hmahmood) [inline pkg/runtimecompiler/config/config.go:29]: Should the configs https://github.com/DataDog/datadog-agent/pull/11760/files#diff-7a70d1343f42219c3b2c07d3bf90635f3d79b9e4fa7ed18ede433218e25e929eR75-R81 move here explicitly?

File: pkg/runtimecompiler/runtimecompiler.go
Comment (hmahmood) [inline pkg/runtimecompiler/runtimecompiler.go:79]: This should be `rc.enabled`? The affect is probably the same since the global looks like a singleton, but I think this would be more correct.

File: cmd/system-probe/app/compile.go
Comment (ISauve) [reply cmd/system-probe/app/compile.go:44]: It would be an init container on the system-probe container, so I believe if the init container fails then the agent container would be okay, but the system-probe container would fail

File: pkg/config/system_probe.go
Comment (ISauve) [reply pkg/config/system_probe.go:171]: `enable_runtime_compliation` is a module-specific setting which only applies to assets within the module - i.e. runtime security has an `enable_runtime_compliation` setting which enables compilation of the runtime security probe & the constant fetcher, and NPM has an `enable_runtime_compliation` setting which enables compilation of the network tracer, http tracer and ebpf conntracker.
`skip_runtime_compilation` skips the entire runtime compilation module, so it doesn't matter what other settings are enabled, runtime compilation will not occur when the system-probe is run.
The purpose of this combination of settings is that you can configure specifically which assets you want to enable runtime compilation for,  set `skip_runtime_compilation` to `true` and run the system-probe's `compile` command (i.e. in an init container) and compilation will occur as expected. Then, when you run the system-probe afterwards, runtime compilation will not occur, but the assets which you have enabled runtime compilation for will try and load the runtime compiled `.o` files

File: pkg/runtimecompiler/runtimecompiler.go
Comment (ISauve) [reply pkg/runtimecompiler/runtimecompiler.go:79]: yep, you're right! same goes for the other 2 things below it

File: cmd/system-probe/modules/compile.go
Comment (ISauve) [reply cmd/system-probe/modules/compile.go:59]: I didn't implement this because there wasn't a reason to - though I suppose it would be useful for if we want to include runtime compilation results in flares/in the `status` command output. That change may be better in another PR though, since this one is already pretty large

File: pkg/runtimecompiler/config/config.go
Comment (hmahmood) [reply pkg/runtimecompiler/config/config.go:29]: So we have `ebpf.Config` nested multiple times here? i.e. it is part of `runtimecompiler.Config`, and e.g., `network.Config` which itself is nested in `runtimecompiler.Config`? 

I think it would be much clearer if we had another struct for the runtime compiler settings that are currently in `ebpf.Config` and that gets included in `runtimecompiler.Config`. I would also recommend you not include `network.Config` and `security.Config` in `runtimecompiler.Config`, and see if you could just have fields that are needed (not the whole structs). I would prefer you not use `ebpf.Config` at all here for these settings.

File: cmd/system-probe/app/compile.go
Comment (hmahmood) [reply cmd/system-probe/app/compile.go:44]: init containers are pod-level, so in a k8s environment, that would run first and then the other containers in the pod will run.

File: pkg/config/system_probe.go
Comment (hmahmood) [reply pkg/config/system_probe.go:171]: That makes sense, but the naming is a little confusing. But I don't have a better suggestion other than `runtime_compilation_config.enabled` (`false` for backwards compatibility?)

Comment (ISauve) [reply pkg/config/system_probe.go:171]: I wanted to stay away from that because if you set `runtime_compilation.enabled` to `false`, IMO the expectation would be that no runtime-compiled asset is going to be loaded - whereas what it _really_ means is that runtime compilation is still enabled, and assets which have been runtime compiled are going to be loaded, but the runtime compilation module which performs the compilation itself is going to be skipped

File: cmd/system-probe/app/compile.go
Comment (ISauve) [reply cmd/system-probe/app/compile.go:44]: ah, yep you're right. An error in that container would prevent all the other ones from starting.

File: pkg/runtimecompiler/config/config.go
Comment (ISauve) [reply pkg/runtimecompiler/config/config.go:29]: What you're suggesting doesn't solve the main issue though - which is that the network tracer, security tracer, OOM kill probe & TCP queue length probe all need access to the runtime compiler settings for calls like this: https://github.com/DataDog/datadog-agent/blob/c8bbffd3ae0314ef30f4cce33d10c31927b262eb/pkg/network/tracer/connection/kprobe/tracer.go#L89

And since the OOM Kill probe and TCP queue length probes receive `ebpf.Config` as their main config, no matter how you slice it the runtime compiler settings will need to visible in some way from within `ebpf.Config`: https://github.com/DataDog/datadog-agent/blob/c8bbffd3ae0314ef30f4cce33d10c31927b262eb/pkg/collector/corechecks/ebpf/probe/oom_kill.go#L39-L40

We could have the runtime compiler settings exist within their own config which gets nested in `ebpf.Config`, but then from `network.Config` and `security.Config` we would have to access those settings through 2 layers of nesting. I don't really think that would be any clearer, especially since the runtime compilation settings currently exist within `ebpf.Config`, so it's a natural place for them to be.

On a different note - I can definitely remove `network.Config` and `security.Config` from `runtimecompiler.Config`, which also means that `ebpf.Config` wouldn't be nested multiple times. The only reason I had opted to have both those configurations in their entirety be fields in `runtimecompiler.Config` was so that when people add new runtime compilation assets in the future, they won't need to add settings for those assets in 2 separate places. But that's also not something I feel strongly about, so if you think it would be clearer to remove those fields I'm fine with that

File: pkg/config/system_probe.go
Comment (hmahmood) [reply pkg/config/system_probe.go:171]: Then something like `runtime_compilation_config.skip_compilation` would be better imo.

File: pkg/runtimecompiler/config/config.go
Comment (hmahmood) [reply pkg/runtimecompiler/config/config.go:29]: It sounds like there are two sets of configs -- one that is global and another that is module specific (for network/security, etc.). There is a singleton instance of the runtime module, so one option could be to expose the runtime compiler config as an exported property of that singleton for other modules to consume. 

I think it would be clearer to not nest the network and security configs -- they can be pretty big as well as muddling the context a little as well.

Comment (ISauve) [reply pkg/runtimecompiler/config/config.go:29]: That would be a perfect solution.... except for this annoying piece of code: https://github.com/DataDog/datadog-agent/blob/c8bbffd3ae0314ef30f4cce33d10c31927b262eb/pkg/runtimecompiler/runtimecompiler.go#L176-L178

Basically, the security probe's runtime-compiled constant fetcher dynamically generates c code which needs to be compiled, so we need to generate that code ourselves in order to compile it before the security probe runs (and then the security probe will later generate it again in order to hash the input & load the correct runtime-compiled output file). This creates a dependency between the compiler and the security probe/the security probe's `constantfetch` module, which means that we can't also have the inverse dependency where the security probe & constant fetcher need to read a config setting from the compiler.

The only way around this would be to have the functions related to generating the C code live inside their own module. This is possible, however these functions are very closely tied to the `constantfetch` and `probe` modules of the security probe, so this extraction makes those modules significantly less clear.

Comment (hmahmood) [reply pkg/runtimecompiler/config/config.go:29]: It only seems to be using one config: https://github.com/DataDog/datadog-agent/blob/c8bbffd3ae0314ef30f4cce33d10c31927b262eb/pkg/security/probe/constantfetch/runtime_compiled.go#L109? I think you can make some modifications there to accommodate it reading from a global config object that has that config.

Comment (ISauve) [reply pkg/runtimecompiler/config/config.go:29]: Yep, `RuntimeCompilerOutputDir` is the only config setting which any of this applies to - I just didn't want to treat it differently from the other runtime compilation settings because I thought it was neater to keep all them all in the same place. I don't feel strongly about that though, so I'll just go ahead and move the other settings to `runtimecompiler.Config` to keep this discussion simpler

Comment (ISauve) [reply pkg/runtimecompiler/config/config.go:29]: Now, `RuntimeCompilerOutputDir` is the only runtime compilation related setting left in `ebpf.Config`. Since all the modules need to access this setting, our options are to leave it where it is, or to make a global variable (/global configuration structure) to hold it. Since runtime compilation is related to ebpf, I don't personally see an issue with runtime compilation settings living within the ebpf settings. For this reason, I don't think there's a strong enough case here to make it a global variable.

(FWIW, all of this discussion has no impact on customers, as either way these settings will all live under `runtime_compilation_config` in the `system-probe.yaml` file.)

Comment (hmahmood) [reply pkg/runtimecompiler/config/config.go:29]: The new changes look great btw! 

I tend to see it as being left out there on its own so I would personally wonder why that is there and everything else in `runtimecompiler.Config`. What problems do you see with something like `runtimecompiler.RuntimeCompiler.Config` being used by the other modules to access all these settings?

Comment (ISauve) [reply pkg/runtimecompiler/config/config.go:29]: I don't personally see a way to get either the `constantfetch` or `probe` package to be able to read from `runtimecompiler.RuntimeCompiler.Config` neatly, because of the dependency cycle I mentioned above. Removing the dependency cycle would require restructuring the constant fetcher code in a way which I think would make it significantly less clear.

We could have all the other modules read from `runtimecompiler.RuntimeCompiler.Config` and do something different in the security module - specifically, we could pass `RuntimeCompilerOutputDir` to the security probe upon initialization and store the value within the probe itself. I don't personally like this solution very much, but regardless, my bigger concern with this pattern is that if we (or any other module) want to do the same thing as the security module is currently doing (aka generating code for compilation dyamically), we'll have import cycles to deal with ourselves. I'd prefer to stick to a model where the runtime compiler can access all the other modules if it needs to, and the other modules don't even need to know the compiler exists.

Comment (ISauve) [reply pkg/runtimecompiler/config/config.go:29]: @paulcacheux  That's definitely good to know, thanks!

@hmahmood  I changed the name of `RuntimeCompilerOutputDir` in an effort to make it more clear why it belongs in the eBPF config - hopefully that alleviates your concerns a little

Comment (hmahmood) [reply pkg/runtimecompiler/config/config.go:29]: Sounds like with these sorts of requirements, we are going to have an issue with cyclic dependencies, especially with the global singleton we have for the runtime compiler. The cycle for the config you can likely solve by having a global in the `runtimecompiler/config` package. But it sounds like this will be an issue for other components of the runtime compiler module in the future. 

I'll leave it to you to decide if you want to move to using a global from the config package instead. For now, I think this'll do. Thanks for making all those changes!

**Review by hmahmood — APPROVED** [review]

File: cmd/system-probe/app/compile.go
Comment (brycekahle) [reply cmd/system-probe/app/compile.go:44]: Do we know how long it takes to compile everything? I worry about adding a delay to all the other containers starting.

File: cmd/system-probe/modules/compile.go
Comment (brycekahle) [inline cmd/system-probe/modules/compile.go:29]: What do you think about each module having a `CompileFn` separate from the `Fn`? That way on the `compile` command, the loader just runs each module's `CompileFn`. I think this might also allow a better separation of concerns with the compiler. I don't like how it currently has to know about all the modules, their configs, and their compilation needs in the same place.

File: cmd/system-probe/config/config.go
Comment (brycekahle) [inline cmd/system-probe/config/config.go:208]: We actually need to know the status NPM separate from USM. This allows parts of the NPM backend to drop data once it is no longer needed.

File: pkg/runtimecompiler/config/config.go
Comment (ISauve) [reply pkg/runtimecompiler/config/config.go:29]: Sounds good. Thanks a bunch for the review Hasan!

**Review by brycekahle — CHANGES_REQUESTED** [review]
We need to try the alternative architecture we chatted about before merging this.

**Comment by ISauve** [comment]
Closing this as we've decided not to move forward with this approach. Certain parts which we do want to keep have been extracted into #13527 & #13568.

**Pull request closed without merging**
